pageserver/src/layered_repository/layer_map.rs

```diff
@@ -340,16 +236,114 @@ impl LayerMap {
     }
 }
 
-impl Default for LayerMap {
+///
+/// Per-segment entry in the LayerMap::segs hash map. Holds all the layers
+/// associated with the segment.
+///
+/// The last layer that is open for writes is always an InMemoryLayer,
+/// and is kept in a separate field, because there can be only one for
+/// each segment. The older layers, stored on disk, are kept in a
+/// BTreeMap keyed by the layer's start LSN.
+struct SegEntry {
+    pub open: Option<Arc<InMemoryLayer>>,
+    pub historic: BTreeMap<Lsn, Arc<dyn Layer>>,
+}
+
+impl Default for SegEntry {
     fn default() -> Self {
-        LayerMap {
-            segs: HashMap::new(),
-            open_segs: BinaryHeap::new(),
-            current_generation: 0,
+        SegEntry {
+            open: None,
+            historic: BTreeMap::new(),
         }
     }
 }
 
+impl SegEntry {
+    pub fn get(&self, lsn: Lsn) -> Option<Arc<dyn Layer>> {
+        if let Some(open) = &self.open {
+            if open.get_start_lsn() <= lsn {
+                let x: Arc<dyn Layer> = Arc::clone(open) as _;
+                return Some(x);
+            }
+        }
+
+        if let Some((_start_lsn, layer)) = self
+            .historic
+            .range((Included(Lsn(0)), Included(lsn)))
```

Comment:
Can we do `..=lsn`

Reply:
I was a bit reluctant to do that, it looks very terse, and I'm used to the syntax. But after staring at it for a few minutes and getting used to it, I agree it's better that way. Changed.

Reply:
Yeah, I get what you're saying. If there's a way to be more explicit while leaving out the other bound at min or max I think that'd be ok too